### ISSUE TITLE
Cannot build as a framework

### ISSUE THREAD

**Issue by jhs** [cloudant/CDTDatastore#194]
(I am working from one week of total iOS, and XCode experience, so please bear with me!)

I think this project cannot build as a framework when used by a Swift project (at least with XCode 7 beta 6). The main problem is "Include of non-modular header inside framework module".

From my limited research, I gather that this issue is somewhat new (as apparently frameworks are not the only way to organize subprojects); but it is becoming more urgent, since Swift projects cannot work around it. (One can disable this compiler error; however, the disabling is not honored in Swift projects, whether by bug or design.) A clear description of the problem is here: http://inessential.com/2015/03/16/swift_blocker

In any case, it sounds like the situation XCode caught is indeed problematic and should be fixed. (My understanding is that, .h files should not import system-wide .h files--.m files should; something about a naming conflict as multiple frameworks compile together). Here is your dependency, CocoaLumberjack [struggling with the same issue][lumberjack-issue].

[lumberjack-issue]: https://github.com/CocoaLumberjack/CocoaLumberjack/issues/544

So, modulo my naïve understanding, to be complete (except of course as a NOFIX), CDTDatastore needs to fix its own issues and then also upgrade to all dependencies when they fix theirs. This issue only applies to this project's specific build failure.

AFAICT, there is only one problem in the codebase:

    /Users/jhs/src/cloudant/CDTDatastore/Classes/common/Encryption/Attachments/CDTBlobEncryptedDataConstants.h:20:9: Include of non-modular header inside framework module 'CDTDatastore.CDTBlobEncryptedDataConstants'

**Comment by rhyshort** [comment]
Unfortunately FMDB, also lacks support for frameworks, although that is being worked on see: https://github.com/ccgus/fmdb/issues/309 .There is a way around the frameworks issue. You need to add a bridging header to your project and use that to import CDTDatastore for use in your swift code.

The easiest way I've found to create a bridging header, is to create a new cocoa class from in the new file wizard, changing the language to objective-c and selecting the option to create the bridging header. In the bridging header you then add the line `#import "CloudantSync.h"`. You'll have access to all the CDTDatastore classes from your swift code automatically, no need for `import` statements in your swift files.

**Comment by jhs** [comment]
Thanks @rhyshort! 

Yes, my larger mission was, in fact, to work without the bridging header. That seems to be the future, AFAIK. CocoaPods tells Swift users to `use_frameworks!` and the GitHub issues on all these projects indicates that we are in a transitional period.

For expediency, I will work with a bridging header, thanks!

Again, I am a complete novice so I haven't got the perspective for this project or any Mac/iOS development. But my larger question is, is framework support an eventual goal for CDTDatastore?

**Comment by rhyshort** [comment]
Yes, we plan to support frameworks as soon as possible, once https://github.com/ccgus/fmdb/issues/309
get's fixed, we'll be looking into making CDTDatastore build as framework again.

**Comment by phatblat** [comment]
The hack mentioned in ccgus/fmdb#309 is enough to work around the FMDB issues with building that module. However, once CDTDatastore is installed as a framework, my app's `import CDTDatastore` line generates the following error: `Include of non-modular header inside framework module 'CDTDatastore.CDTBlobEncryptedDataConstants'`

This error is coming from the following import statement:

```objc
#import <CommonCrypto/CommonCryptor.h>
```

**Comment by rhyshort** [comment]
Thanks for find that, I'll look into it when fmdb officially supports building a framework.

**Comment by mikerhodes** [comment]
Hmm that's a bummer, I'd've hoped that CommonCrypto was modularised. Perhaps we buggered something up when building that bit.

**Comment by simonmcl** [comment]
I've just used version 1.0.0 with `use_frameworks!` on Xcode 7.2.1 and its not working. Is there something else required?  trying to remove `use_frameworks!` and add a bridging header is causing us a lot of issues with other pods.

**Comment by rhyshort** [comment]
Framework support hasn't been released yet, the current plan is to have it included in 1.1, but there are some other strategic priorities we'd like to include in the that release as well.

**Comment by suculent** [comment]
I'm just disabling Defines Module and Allowing Non-Modular frameworks to include CDT in my otherwise Swift-only project.

**Comment by simonmcl** [comment]
@rhyshort ok thanks. Is there any roadmap for 1.1? Were looking to integrate this into some products, but hacking are way around it isn't something we want to do. We got around it for our next POC by moving another pod to Objective-C, not a great story to tell. I would have figured swift interoperability would have been on the top of the list given IBM's shift.

**Comment by simonmcl** [comment]
Has there been any update on this?

Seems someone created a "SwiftFramework" branch inside FMDB 4 months ago, its not merged into master but may still solve the issue: https://github.com/ccgus/fmdb/tree/swiftFramework

Also why is this issue closed? isn't this still an issue? You are blocked on a dependency and have a plan to fix it in 1.1, but it hasn't been resolved yet. So should it not remain open until there is a fix available in a cocoapod release?

**Comment by rhyshort** [comment]
The issue is closed because CDTDatastore on Master can successfully built as a framework using the CocoaPods `use_frameworks!` directive.  FMDB fixed the issue of building as a framework in 2.6, which is the version we use. Our sample project makes use of this see [Project/Podfile](https://github.com/cloudant/CDTDatastore/blob/master/Project/Podfile).


The issue is closed because it is resolved, our work flow is to close issues when the code is complete as this allows us to better track our status moving towards the next release.  Keeping the issue in the open state would make it difficult to track what we have done, and what issues we have yet to resolve which may be blocking the release. Therefore we attempt to close issues when the fix has been merged into master, which indicates it is ready for release.
